### PR TITLE
[Merged by Bors] - feat(ModelTheory): Fraïssé limits are unique

### PR DESCRIPTION
**Pull request by GabinKolly**
Prove that Fraïsse limits are unique up to equivalence.

Co-authored-by: Aaron Anderson <awainverse@gmail.com>

---

- [x] depends on: #9953
- [x] depends on: #11175

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by awainverse** [comment]
Can you merge master back into this? The diff is too large, and is counting the previous PR as new additions.

**Review by awainverse — COMMENTED** [review]
All of this looks good to me, but it would probably be easier to review in greater detail if it is split into shorter PRs. I would recommend the following structure:

- The edits to `Substructures.lean` and `DirectLimit.lean` that don't refer to `SubEquivalence`s
- The definition of `SubEquivalence` in `Substructures.lean`
- The work with `SubEquivalence`s in `DirectLimit.lean` and the conclusion in `Fraisse.lean`

with other minor edits to other files put in where they're needed.

**Comment by YaelDillies** [comment]
I have taken the freedom to mark this PR as depending on #11175 since you claim this is the case in the PR description of #11175. It was a bit confusing seeing a PR seemingly ready, but with merge conflicts and no dependencies 😅

If you need help with #11175, please let us know and we will help!

**Comment by awainverse** [comment]
Since this has been `awaiting-author` for some time, and I just remembered that I am credited as co-author, I'm going to start editing this directly.

File: Mathlib/ModelTheory/testing.lean
Comment (YaelDillies) [inline Mathlib/ModelTheory/testing.lean:1]: Can this file be in the `test` folder instead?

**Comment by YaelDillies** [comment]
Nice! It looks much easier to review than the 2k lines monster it previously was. Is the PR title still accurate?

File: Mathlib/ModelTheory/Substructures.lean [hunk 1]
```diff
@@ -341,14 +342,29 @@ theorem closure_univ : closure L (univ : Set M) = ⊤ :=
...
   haveI : Small.{u} (∅ : Set M) := small_subsingleton _
   exact Substructure.small_closure
 
+theorem iSup_eq_closure {ι : Sort*} (S : ι → L.Substructure M) :
+    ⨆ i, S i = closure L (⋃ i, (S i : Set M)) := by simp_rw [closure_iUnion, closure_eq]
+
+-- This proof uses the fact that `Substructure.closure` is finitary.
+theorem mem_iSup_of_directed {ι : Type*} [hι : Nonempty ι] {S : ι → L.Substructure M}
+    (hS : Directed (· ≤ ·) S) {x : M} :
+    x ∈ (iSup S : L.Substructure M) ↔ ∃ i, x ∈ S i := by
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Substructures.lean:359]:
```suggestion
    x ∈ ⨆ i, S i ↔ ∃ i, x ∈ S i := by
```

File: Mathlib/ModelTheory/Fraisse.lean [hunk 2]
```diff
@@ -302,6 +365,56 @@ theorem isFraisse [Countable (Σ l, L.Functions l)] [Countable M] (h : IsFraisse
     IsFraisse K :=
   (congr rfl h.age).mp h.ultrahomogeneous.age_isFraisse
 
+variable {K} {N : Type w} [L.Structure N]
+variable [Countable (Σ l, L.Functions l)] [Countable M] [Countable N]
+variable (hM : IsFraisseLimit K M) (hN : IsFraisseLimit K N)
+
+include hM hN
+
+protected theorem IsExtensionPair : L.IsExtensionPair M N := by
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Fraisse.lean:374]:
```suggestion
protected theorem isExtensionPair : L.IsExtensionPair M N := by
```

File: Mathlib/ModelTheory/Fraisse.lean [hunk 3]
```diff
@@ -302,6 +365,56 @@ theorem isFraisse [Countable (Σ l, L.Functions l)] [Countable M] (h : IsFraisse
...
+  intro ⟨f, f_FG⟩ m
+  let S := f.dom ⊔ closure L {m}
+  have dom_le_S : f.dom ≤ S := by
+    simp only [S, closure_union, closure_eq, ge_iff_le, le_sup_left]
+  have S_FG : S.FG := f_FG.sup (Substructure.fg_closure_singleton _)
+  have S_in_age_N : ⟨S, inferInstance⟩ ∈ L.age N := by
+    rw [hN.age, ← hM.age]
+    exact ⟨(fg_iff_structure_fg S).1 S_FG, ⟨subtype _⟩⟩
+  let nonempty_S_N : Nonempty (S ↪[L] N) := by
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Fraisse.lean:383]:
```suggestion
  have nonempty_S_N : Nonempty (S ↪[L] N) := by
```

File: Mathlib/ModelTheory/Fraisse.lean [hunk 4]
```diff
@@ -302,6 +365,56 @@ theorem isFraisse [Countable (Σ l, L.Functions l)] [Countable M] (h : IsFraisse
...
+  let ⟨g, eq⟩ := hN.ultrahomogeneous.extend_embedding (f.dom.fg_iff_structure_fg.1 f_FG)
+    ((subtype f.cod).comp f.toEquiv.toEmbedding) (inclusion dom_le_S)
+  refine ⟨⟨⟨S, g.toHom.range, g.equivRange⟩, S_FG⟩,
+    subset_closure.trans (le_sup_right : _ ≤ S) (mem_singleton m), ?_⟩
+  simp only [Subtype.mk_le_mk, PartialEquiv.le_def]
+  use dom_le_S
+  rw [eq]
+  rfl
+
+theorem unique_FraisseLimit : Nonempty (M ≃[L] N) := by
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Fraisse.lean:395]:
```suggestion
theorem nonempty_FraisseLimit : Nonempty (M ≃[L] N) := by
```
also what is the `FraisseLimit` referring to?

File: Mathlib/ModelTheory/Substructures.lean (hunk 1 above)
Comment (awainverse) [reply Mathlib/ModelTheory/Substructures.lean:359]: I found a few more of these in the file - taking the liberty to change them now.

File: Mathlib/ModelTheory/Fraisse.lean (hunk 4 above)
Comment (awainverse) [reply Mathlib/ModelTheory/Fraisse.lean:395]: I've added a docstring for this important theorem (although it also was in the module doc), and I'm changing the name.
It's already in the `IsFraisseLimit` namespace, so `IsFraisseLimit.nonempty_equiv` now proves that any two Fraïssé limits (`M` and `N`) admit a (not necessarily unique) isomorphism.

Comment (YaelDillies) [reply Mathlib/ModelTheory/Fraisse.lean]: Oh yeah, that makes more sense. Sorry for being slow

File: Mathlib/ModelTheory/Fraisse.lean [hunk 5]
```diff
@@ -265,6 +283,51 @@ structure IsFraisseLimit [Countable (Σ l, L.Functions l)] [Countable M] : Prop
 
 variable {M}
 
+/-- Any embedding from a finitely generated `S` to an ultrahomogeneous structure `M`
+can be extended to an embedding from any structure with an embedding to `M`.-/
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Fraisse.lean:287]:
```suggestion
can be extended to an embedding from any structure with an embedding to `M`. -/
```

File: Mathlib/ModelTheory/Fraisse.lean [hunk 6]
```diff
@@ -265,6 +283,51 @@ structure IsFraisseLimit [Countable (Σ l, L.Functions l)] [Countable M] : Prop
...
+theorem IsUltrahomogeneous.extend_embedding (M_homog : L.IsUltrahomogeneous M) {S : Type*}
+    [L.Structure S] (S_FG : FG L S) {T : Type*} [L.Structure T] [h : Nonempty (T ↪[L] M)]
+    (f : S ↪[L] M) (g : S ↪[L] T) :
+    ∃ f' : T ↪[L] M, f = f'.comp g := by
+  let ⟨r⟩ := h
+  let s := (r.comp g)
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Fraisse.lean:293]:
```suggestion
  let s := r.comp g
```

File: Mathlib/ModelTheory/Substructures.lean [hunk 7]
```diff
@@ -341,14 +342,29 @@ theorem closure_univ : closure L (univ : Set M) = ⊤ :=
...
   haveI : Small.{u} (∅ : Set M) := small_subsingleton _
   exact Substructure.small_closure
 
+theorem iSup_eq_closure {ι : Sort*} (S : ι → L.Substructure M) :
+    ⨆ i, S i = closure L (⋃ i, (S i : Set M)) := by simp_rw [closure_iUnion, closure_eq]
+
+-- This proof uses the fact that `Substructure.closure` is finitary.
+theorem mem_iSup_of_directed {ι : Type*} [hι : Nonempty ι] {S : ι → L.Substructure M}
+    (hS : Directed (· ≤ ·) S) {x : M} :
+    x ∈ ⨆ i, S i ↔ ∃ i, x ∈ S i := by
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Substructures.lean:359]: Can we have the `sSup` version of this lemma?

**Comment by YaelDillies** [comment]
I updated the PR description in a similar way to the PR title.

maintainer merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**